### ISSUE TITLE
Support network-namespace based wireguard vpn setup [feature request]

### ISSUE THREAD

**Issue by anderspapitto** [NixOS/nixpkgs#52411]
At https://www.wireguard.com/netns/, the wireguard documentation recommends as the top option a vpn setup based on a network user-namespace, rather than routing. It would probably require some integration with other services - for example, variants of the wpa_supplicant and dhcpcd services that run inside the network namespace.

**Comment by anderspapitto** [comment]
So, to brainstorm what this might consist of

(1) a new systemd service `wg0-namespace`, which 
(a) creates a network namespace named `physical`
(b) moves `eth0` and `wlan0` (or equivalent) into that namespace
(c) creates the wireguard interface `wg0` inside namespace `physical`
(d) moves `wg0` to the `init` namespace
`phys-net-ns` would conflict with `wpa_supplicant` and `dhcpcd`as they would no longer have access to `eth0`/`wlan0`.
(2) two new systemd services, `wg0-supplicant` and `wg0-dhcpcd`, which are equivalent to the `wpa_supplicant` and `dhcpcd` services, respectively, except that they wrap ExecStart with `ip netns exec physical ...`
these two services depend on `wg0-namespace`
(3) a new systemd service, `wg0-vpn`, which (all in the init namespace)
(a) configures the wg0 interface with private key, peer IP and so on
(b) brings up the interface
(c) adds the interface as default route
(4) optionally a cli utility with the `physexec` functionality described in the wireguard docs, to allow explicitly launching process to bypass the vpn

possibly (3) should just be folded into (1).

ideally shutting off the vpn would automatically restore `wpa_supplicant` and `dhcpcd` services (in the default root namespace). I'm not sure if this can be expressed explicitly with systemd units - we might just need an explicit `systemctl` invocation in ExecStopPost

**Comment by anderspapitto** [comment]
I have a prototype implementation here https://github.com/anderspapitto/nixos-configuration/commit/ff73f584e717e84ccab11a454d5f4987082e2ee3, with one small tweak to nixpkgs needed here https://github.com/anderspapitto/nixpkgs/commit/3429be383a98a5213ec1b509a5145d3d23a8d7a7

It works great, but some parts are a bit hacky at the moment. What would really help is a clean way to say "restart this systemd service inside this user namespace". I'm not sure if there's any support for this type of thing in systemd itself, or in nixos. One thought is to maybe use nixos' containers, if they support namespaces.

**Comment by anderspapitto** [comment]
my implementation is improved at https://github.com/anderspapitto/nixos-configuration/blob/master/wireguard-client.nix, and discussed at https://reflexivereflection.com/posts/2018-12-18-wireguard-vpn-with-network-namespace-on-nixos.html. That's probably where I'll let things rest, as it's fully functional for my use case.

**Comment by lheckemann** [comment]
Thanks for the write-up! I'll probably be stealing this ;)

**Comment by eadwu** [comment]
Just a question, but where does `enp0s25` come from? I see that the wireless interface is included at `networking.wireless.interfaces` but I'm lost on where `enp0s25` comes from. I only have a single network interface, `wlp59s0`.

**Comment by eadwu** [comment]
I managed to configure the netns after it got exposed to `ip netns` but `JoinsNamespaceOf` and `PrivateNetwork` still require the use of `ip netns exec` to actually use it (or maybe the way I'm setting it up is wrong)

`ls -la /sys/class/net`
```
fg12-unit-script-netns-test-start[30310]: drwxr-xr-x  2 root root 0 Dec 31 10:51 .
fg12-unit-script-netns-test-start[30310]: drwxr-xr-x 55 root root 0 Dec 31 10:51 ..
fg12-unit-script-netns-test-start[30310]: lrwxrwxrwx  1 root root 0 Dec 31 10:51 docker.
fg12-unit-script-netns-test-start[30310]: lrwxrwxrwx  1 root root 0 Dec 31 10:51 lo
fg12-unit-script-netns-test-start[30310]: lrwxrwxrwx  1 root root 0 Dec 31 10:51 wlp...
```

`ip netns exec physical ls -la /sys/class/net`
```
fg12-unit-script-netns-test-start[30310]: drwxr-xr-x  2 root root 0 Dec 31 10:51 .
fg12-unit-script-netns-test-start[30310]: drwxr-xr-x 55 root root 0 Dec 31 10:51 ..
fg12-unit-script-netns-test-start[30310]: lrwxrwxrwx  1 root root 0 Dec 31 10:51 lo
```

My configuration is
```nix
systemd.services = {
    "netns-${NETNS}" = {
      # Automatically create namespace
      # after = [ "run-netns.mount" ];
      # wants = [ "run-netns.mount" ];
      # wantedBy = [ "network.target" ];

      description = "Wireguard network namespace `${NETNS}`";
      documentation = [ "https://github.com/systemd/systemd/issues/2741#issuecomment-433979748" ];
      path = with pkgs; [ iproute utillinux ];
      serviceConfig = {
        Type = "oneshot";
        PrivateNetwork = true;
        RemainAfterExit = true;
      };

      postStop = ''
        ip netns del ${NETNS}
      '';

      script = ''
        ip netns add ${NETNS}
        umount /run/netns/${NETNS}
        mount --bind /proc/self/ns/net /run/netns/${NETNS}
      '';
    };

    netns-test = {
      # after = [ "netns-${NETNS}" ];
      # bindsTo = [ "netns-${NETNS}" ];
      # requires = [ "netns-${NETNS}" ];

      path = with pkgs; [ iproute ];
      unitConfig.JoinsNamespaceOf = "netns-${NETNS}.service";
      serviceConfig = {
        PrivateNetwork = true;
      };
      script = ''
        _NETNS=$(ip netns identify)

        ip netns list
        echo "CURRENT NETNS"
        echo "netns ''${_NETNS}"
        ip netns identify
        ls -la /sys/class/net
        ip netns exec ${NETNS} ls -la /sys/class/net
        echo "END"
      '';
    };
};
```

**Comment by eadwu** [comment]
The problem of `ip link` for me is that it doesn't only output only the wireless interfaces (even though for me the wl* matches all of my wireless interfaces).

**Comment by orivej** [comment]
I think it is easier to leave physical interfaces in the initial network namespace: then NetworkManager just works.

**Comment by eadwu** [comment]
Based on the article, it looks like interfaces have to be in the same network namespace in order for all the traffic to go through wg0.

**Comment by orivej** [comment]
The WireGuard interfaces should be created in the namespace with internet connectivity (e.g. the namespace with physical interfaces), and then moved into the namespace with no interfaces (other than loopback). The [netns article](https://www.wireguard.com/netns/) explains how to create a new namespace, move physical interfaces there, reconfigure them (because they loose their settings when they are moved), create a wireguard interface in that new namespace, and then move it into the original namespace. The benefit of this scheme is that it makes wireguard available to all applications running in the original namespace, but the downside is the ensuing difficulty of configuring the physical interfaces after the move. I have shown a NixOS config that avoids this difficulty, but requires you to list the desktop and other services that you want to run other the VPN.

**Comment by tpanum** [comment]
For people seeking a solution to the `Allowed IPs = 0.0.0.0/0` problem, the newer `networking.wg-quick.interfaces` seems to be working flawlessly.

**Comment by matthias-t** [comment]
see also: https://discourse.nixos.org/t/run-systemd-service-in-network-namespace/

**Comment by flokli** [comment]
FYI: The whole "creating the namespace, and moving the wireguard interface into the namespace" part should be much easier once https://github.com/systemd/systemd/pull/14915 has landed.

**Comment by Zhen-hao** [comment]
> see also: https://discourse.nixos.org/t/run-systemd-service-in-network-namespace/

@matthias-t  can you please be more explicit how to use your solution to routing all traffic via Wireguard? I tried to follow your blog but it seems I cannot connect to the `wg` service which runs Wireguard.

I don't know if the following part is relevant in this case.
```
{ ... }: {
  systemd.services.<service> = {
    bindsTo = [ "wg.service" ];
    after = [ "wg.service" ];
    unitConfig.JoinsNamespaceOf = "netns@wg.service";
    serviceConfig.PrivateNetwork = true;
  };
}
```

**Comment by Zhen-hao** [comment]
> For people seeking a solution to the `Allowed IPs = 0.0.0.0/0` problem, the newer `networking.wg-quick.interfaces` seems to be working flawlessly.

@tpanum  do you mean it works on the client-side or the server-side?

**Comment by matthias-t** [comment]
@Zhen-hao I'm not sure what you mean by "routing all traffic via Wireguard". The point of my post was to confine specific systemd services, not routing traffic for all processes. The service I called `wg` configures a wireguard interface in the `wg` netns. Its purpose is to isolate systemd services with `JoinsNamespaceOf`. But you can also just `sudo ip netns exec wg sudo -u <user> bash` for a shell.

**Comment by Zhen-hao** [comment]
@matthias-t sorry for the confusion. I thought you meant your solution could also solve the "Allowed IPs = 0.0.0.0/0" issue mentioned earlier in this thread. That's what I called "routing all traffic"

Also, you blog looks similar to [this other blog](https://reflexivereflection.com/posts/2018-12-18-wireguard-vpn-with-network-namespace-on-nixos.html). Do you think your solution can be used to achieve the same goal?

**Comment by tpanum** [comment]
@Zhen-hao I believe many people get here by wanting to use Wireguard as a VPN, by basically routing all traffic through some service that support Wireguard (like Mullvad) in order to not disclose their traffic to some untrusted NAT.

What I described have been working great for me, as a client.

**Comment by Zhen-hao** [comment]
@tpanum thank you! I had some trouble on the server-side because I forgot to reboot after setting `"net.ipv4.ip_forward" = 1;`

**Comment by yash-ahir** [comment]
> @tpanum thank you! I had some trouble on the server-side because I forgot to reboot after setting `"net.ipv4.ip_forward" = 1;`

Can you elaborate? Did you set that on the server or the client? And what is the exact option? I've been getting the same problem with not being able to ping any other interfaces when setting `ips = 0.0.0.0/0` on the serverside.

**Comment by Zhen-hao** [comment]
"on the server-side"

**Comment by yash-ahir** [comment]
> "on the server-side"

Alright so, using `networking.wireguard` with `"net.ipv4.ip_forward" = 1;` still doesn't seem to work for me, `ips = [ "0.0.0.0/0" ];` for `wg0`. Can you perhaps share your `configuration.nix`?

**Comment by Zhen-hao** [comment]
@yash-ahir 
I actually use something else
```nix
networking.wg-quick.interfaces = {
    # "wg0" is the network interface name. You can name the interface arbitrarily.
    wg0 = {
      # Determines the IP address and subnet of the server's end of the tunnel interface.
      address = [ 
        "10.100.0.1/24"
        ];

      # The port that Wireguard listens to. Must be accessible by the client.
      listenPort = 51820;

      postUp = ''
        ${pkgs.iptables}/bin/iptables -t nat -A POSTROUTING -s 10.100.0.0/24 -o enp3s0 -j MASQUERADE

        # ${pkgs.iptables}/bin/iptables -A FORWARD -i %i -j ACCEPT
        # ${pkgs.iptables}/bin/iptables -A FORWARD -o %i -j ACCEPT
        # ${pkgs.iptables}/bin/iptables -t nat -A POSTROUTING -o enp3s0 -j MASQUERADE
      '';

      # This undoes the above command
      postDown = ''
        ${pkgs.iptables}/bin/iptables -t nat -D POSTROUTING -s 10.100.0.0/24 -o enp3s0 -j MASQUERADE
        
        # ${pkgs.iptables}/bin/iptables -D FORWARD -i %i -j ACCEPT
        # ${pkgs.iptables}/bin/iptables -D FORWARD -o %i -j ACCEPT
        # ${pkgs.iptables}/bin/iptables -t nat -D POSTROUTING -o enp3s0 -j MASQUERADE
      '';

      # Path to the private key file.
      #
      # Note: The private key can also be included inline via the privateKey option,
      # but this makes the private key world-readable; thus, using privateKeyFile is
      # recommended.
      privateKeyFile = "...";

      peers = [
        ...
      ];
    };
  };
```

**Comment by yash-ahir** [comment]
> @yash-ahir
> I actually use something else
> 
> ```nix
> networking.wg-quick.interfaces = {
>     # "wg0" is the network interface name. You can name the interface arbitrarily.
>     wg0 = {
>       # Determines the IP address and subnet of the server's end of the tunnel interface.
>       address = [ 
>         "10.100.0.1/24"
>         ];
> 
>       # The port that Wireguard listens to. Must be accessible by the client.
>       listenPort = 51820;
> 
>       postUp = ''
>         ${pkgs.iptables}/bin/iptables -t nat -A POSTROUTING -s 10.100.0.0/24 -o enp3s0 -j MASQUERADE
> 
>         # ${pkgs.iptables}/bin/iptables -A FORWARD -i %i -j ACCEPT
>         # ${pkgs.iptables}/bin/iptables -A FORWARD -o %i -j ACCEPT
>         # ${pkgs.iptables}/bin/iptables -t nat -A POSTROUTING -o enp3s0 -j MASQUERADE
>       '';
> 
>       # This undoes the above command
>       postDown = ''
>         ${pkgs.iptables}/bin/iptables -t nat -D POSTROUTING -s 10.100.0.0/24 -o enp3s0 -j MASQUERADE
>         
>         # ${pkgs.iptables}/bin/iptables -D FORWARD -i %i -j ACCEPT
>         # ${pkgs.iptables}/bin/iptables -D FORWARD -o %i -j ACCEPT
>         # ${pkgs.iptables}/bin/iptables -t nat -D POSTROUTING -o enp3s0 -j MASQUERADE
>       '';
> 
>       # Path to the private key file.
>       #
>       # Note: The private key can also be included inline via the privateKey option,
>       # but this makes the private key world-readable; thus, using privateKeyFile is
>       # recommended.
>       privateKeyFile = "...";
> 
>       peers = [
>         ...
>       ];
>     };
>   };
> ```

Welp, although it still doesn't work for me, thanks for the help!

**Comment by flokli** [comment]
These kind of debugging sessions should probably better be handled in the forum, or a separate issue.

This issue is about adding support for wireguard with network namespaces, which is somewhat blocked on networkd support for namespaces, as per https://github.com/NixOS/nixpkgs/issues/52411#issuecomment-631985762. We can re-visit this once it has landed.

In the meantime, using wg-quick for such usecases seems to work. Let's close this issue, and open a new one once there's something actionable.

**Comment by benbot** [comment]
Has this ever been resolved?

I'm still getting wireguard connection issues and it looks like #52411 is closed now too.
